### PR TITLE
Avoid redundant eviction disk reads in p23

### PR DESCRIPTION
**Pull request by SirTyson**
# Description

Resolves https://github.com/stellar/stellar-core-internal/issues/402

Avoids redundant disk IO before protocol 24 in the eviction scan.

# Checklist
- [x] Reviewed the [contributing](https://github.com/stellar/stellar-core/blob/master/CONTRIBUTING.md#submitting-changes) document
- [x] Rebased on top of master (no merge commits)
- [x] Ran `clang-format` v8.0.0 (via `make format` or the Visual Studio extension)
- [x] Compiles
- [x] Ran all tests
- [ ] If change impacts performance, include supporting evidence per the [performance document](https://github.com/stellar/stellar-core/blob/master/performance-eval/performance-eval.md)

### CODE REVIEW

File: src/bucket/BucketSnapshot.cpp [hunk 1]
```diff
@@ -314,8 +314,11 @@ LiveBucketSnapshot::scanForEviction(
                 // for persistent entries, we need to make sure we evict the
                 // newest version of the entry, since it will be persisted in
                 // the hot archive. So we add persistent entries to our DB
-                // lookup to find the newest version.
-                if (isPersistentEntry(le.data))
+                // lookup to find the newest version. Note that there was a
+                // bug in protocol 23 where this check was not performed.
+                if (isPersistentEntry(le.data) &&
```
Comment (marta-lokhova) [inline src/bucket/BucketSnapshot.cpp:319]: Can we also skip any scanning if eviction is disabled completely? (per [this](https://github.com/stellar/stellar-core-internal/issues/402#issuecomment-3429421778) comment)

Comment (SirTyson) [reply src/bucket/BucketSnapshot.cpp:319]: I don't think we should. It would add more complexity to the scan logic, and there are only about a week of ledgers where this optimization would be useful, so I don't think it's worth the extra code. There should be no reason to disable eviction in the future (other than bugs)

Comment (dmkozh) [reply src/bucket/BucketSnapshot.cpp:319]: I think disabling the eviction code completely would be useful not as an optimization, but as a bit more defensive approach. As you've mentioned there shouldn't be a reason to disable eviction unless there is a bug, but if there is a bug, then it probably would be safer to not run the eviction code (including the scan) at all. That said, it's only worth doing if the code is straightforward enough, like a simple conditional statement. If it's not as simple, then I think it's indeed not worth the complexity.

Comment (marta-lokhova) [reply src/bucket/BucketSnapshot.cpp:319]: Can you elaborate on code complexity here? Could we  have a fake resolved future that returns an empty list and pretends that the scan was complete?

Comment (SirTyson) [reply src/bucket/BucketSnapshot.cpp:319]: Looking into this a bit more, there is a way to disable eviction scans completely, and that's by setting `evictionScanSize` to 0. The issue is, what we did in p23 was not disable the scan, but just disable the evictions via `maxEntriesToArchive = 0`. This means if a given scan region had no entries to evict, we will do the scan and advance the iterator position. We can't skip the scan in the p23 case since we still actually did the scan and advance the iterator. In retrospect, `evictionScanSize = 0` was the better option, but we were going quickly and it was easier to reason about `maxEntriesToArchive == 0`.

Looking through the scan loop, there shouldn't be any issue with `evictionScanSize == 0` today, which would disable eviction, but still go through much of the scan logic. What I can do is add a high level "kill switch" check that returns early and avoids most of the scan when `evictionScanSize == 0` so we can avoid going through the logic. I'll also add an explicit test case. This won't help in p23 though, since we still have to perform the scan when `maxEntriesToArchive == 0` and can't short cut anything.

**Comment by SirTyson** [comment]
I've added some functionality to exit the scan early if evictionScanSize is 0, and added a test for this case. The test passed both before and after the exit early change. We still need to do some work within the scan no matter what (for example, if the Bucket the iterator is pointing to has changed, the iterator has been invalidated, so we update it even when eviction scans are disabled). This should not be a protocol change.

File: src/bucket/SearchableBucketList.cpp [hunk 2]
```diff
@@ -51,6 +51,15 @@ SearchableLiveBucketListSnapshot::scanForEviction(
     auto startIter = evictionIter;
     auto scanSize = sas.evictionScanSize;
 
+    // exit scan early if there's nothing to scan. We still need to call
+    // `updateStartingEvictionIterator` before exiting early in case the
```
Comment (dmkozh) [inline src/bucket/SearchableBucketList.cpp]: Is there a test that observes this behavior (i.e. iterator updates due to bucket changes)?

Comment (SirTyson) [reply src/bucket/SearchableBucketList.cpp]: We had a test for this when eviction is enabled, "iterator resets when bucket changes". I've extended it to also include the eviction scan size of 0 case. While doing this, I also removed the exit early condition. Even when eviction is disabled, it's important that the eviction iterator remain valid at all times such that we can safely resume eviction when re-enabled. The "shortcut" path actually introduces more complexity, as we skip the scan, but still need to check for bucket spills and potentially re-write the iterator if it's been invalidated. Instead of adding more edge cases, I've removed the early exit and kept the scan as is, but added more tests to ensure we don't actually make progress when eviction scan size is 0, other than maintaining iterator validity.

Comment (dmkozh) [reply src/bucket/SearchableBucketList.cpp]: Alright, yeah, I don't think it's worth writing more complex code to handle the edge cases.

**Review by dmkozh — APPROVED** [review]
